server: TestAddNewStoresToExistingNodes failed

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10811911?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10811911?buildTab=artifacts#/) on master @ [06fde6176f1be7fe29ff04c287771aa223e5d802](https://github.com/cockroachdb/cockroach/commits/06fde6176f1be7fe29ff04c287771aa223e5d802):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 112335 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes1060769749
    test_log_scope.go:81: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29524

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10812861?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10812861?buildTab=artifacts#/) on master @ [1137aee0089fc2fe161bab0f1c899d2ddc997c90](https://github.com/cockroachdb/cockroach/commits/1137aee0089fc2fe161bab0f1c899d2ddc997c90):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 108781 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes3761490612
    test_log_scope.go:81: use -show-logs to present logs inline
    testcluster.go:400: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10812861?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10812861?buildTab=artifacts#/) on master @ [1137aee0089fc2fe161bab0f1c899d2ddc997c90](https://github.com/cockroachdb/cockroach/commits/1137aee0089fc2fe161bab0f1c899d2ddc997c90):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 154859 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes833321938
    test_log_scope.go:81: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10854906?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10854906?buildTab=artifacts#/) on master @ [7cd5444acd8f32ec62aa0077d43263f3087beed0](https://github.com/cockroachdb/cockroach/commits/7cd5444acd8f32ec62aa0077d43263f3087beed0):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 134180 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes875193987
    test_log_scope.go:81: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10855168?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10855168?buildTab=artifacts#/) on master @ [5e8c311598b058181ce05b403966c015e2c3c7a8](https://github.com/cockroachdb/cockroach/commits/5e8c311598b058181ce05b403966c015e2c3c7a8):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 117495 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes3804624714
    test_log_scope.go:81: use -show-logs to present logs inline
    testcluster.go:400: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/multi_store_test.go
+++ b/pkg/server/multi_store_test.go
@@ -12,12 +12,14 @@ package server_test
 
 import (
 	"context"
+	"fmt"
 	"sort"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
@@ -41,79 +43,50 @@ func TestAddNewStoresToExistingNodes(t *testing.T) {
 
 	ctx := context.Background()
 
-	n1s1, n1cleanup1 := testutils.TempDir(t)
-	defer n1cleanup1()
-	n2s1, n2cleanup1 := testutils.TempDir(t)
-	defer n2cleanup1()
-	n3s1, n3cleanup1 := testutils.TempDir(t)
-	defer n3cleanup1()
+	ser := server.NewStickyInMemEnginesRegistry()
+	defer ser.CloseAllStickyInMemEngines()
 
-	numNodes := 3
-	tcArgs := base.TestClusterArgs{
-		ServerArgsPerNode: map[int]base.TestServerArgs{
-			// NB: on my local (beefy) machine, upreplication
-			// takes ~6s. This is pretty hefty compared to ~1s
-			// with ephemeral stores. But - we need the real
-			// stores here. At the time of writing, we perform
-			// ~100 change replicas txns, all in all, and
-			// 0.06s for a replication change does seem ok.
-			0: {StoreSpecs: []base.StoreSpec{{Path: n1s1}}},
-			1: {StoreSpecs: []base.StoreSpec{{Path: n2s1}}},
-			2: {StoreSpecs: []base.StoreSpec{{Path: n3s1}}},
-		},
+	const (
+		numNodes                     = 3
+		numStoresPerNodeInitially    = 1
+		numStoresPerNodeAfterRestart = 3
+	)
+
+	mkClusterArgs := func(numNodes, numStoresPerNode int) base.TestClusterArgs {
+		tcArgs := base.TestClusterArgs{
+			// NB: it's important that this test wait for full replication. Otherwise,
+			// with only a single voter on the range that allocates store IDs, it can
+			// pass erroneously. StartTestCluster already calls it, but we call it
+			// again explicitly.
+			ReplicationMode:   base.ReplicationAuto,
+			ServerArgsPerNode: map[int]base.TestServerArgs{},
+		}
+		for srvIdx := 0; srvIdx < numNodes; srvIdx++ {
+			var serverArgs base.TestServerArgs
+			serverArgs.Knobs.Server = &server.TestingKnobs{StickyEngineRegistry: ser}
+			for storeIdx := 0; storeIdx < numStoresPerNode; storeIdx++ {
+				id := fmt.Sprintf("s%d.%d", srvIdx+1, storeIdx+1)
+				serverArgs.StoreSpecs = append(
+					serverArgs.StoreSpecs,
+					base.StoreSpec{InMemory: true, StickyInMemoryEngineID: id},
+				)
+			}
+			tcArgs.ServerArgsPerNode[srvIdx] = serverArgs
+		}
+		return tcArgs
 	}
 
-	tc := testcluster.StartTestCluster(t, numNodes, tcArgs)
-	// NB: it's important that this test wait for full replication. Otherwise,
-	// with only a single voter on the range that allocates store IDs, it can
-	// pass erroneously. StartTestCluster already calls it, but we call it
-	// again explicitly.
-	if err := tc.WaitForFullReplication(); err != nil {
-		log.Fatalf(ctx, "while waiting for full replication: %v", err)
-	}
+	tc := testcluster.StartTestCluster(t, numNodes, mkClusterArgs(numNodes, numStoresPerNodeInitially))
 	clusterID := tc.Server(0).StorageClusterID()
 	tc.Stopper().Stop(ctx)
 
-	// Add two additional stores to each node.
-	n1s2, n1cleanup2 := testutils.TempDir(t)
-	defer n1cleanup2()
-	n2s2, n2cleanup2 := testutils.TempDir(t)
-	defer n2cleanup2()
-	n3s2, n3cleanup2 := testutils.TempDir(t)
-	defer n3cleanup2()
-
-	n1s3, n1cleanup3 := testutils.TempDir(t)
-	defer n1cleanup3()
-	n2s3, n2cleanup3 := testutils.TempDir(t)
-	defer n2cleanup3()
-	n3s3, n3cleanup3 := testutils.TempDir(t)
-	defer n3cleanup3()
-
-	tcArgs = base.TestClusterArgs{
-		// We need ParallelStart since this is an existing cluster. If
-		// we started sequentially, then the first node would hang forever
-		// waiting for the KV layer to become available, but that only
-		// happens when the second node also starts.
-		ParallelStart:   true,
-		ReplicationMode: base.ReplicationManual, // saves time
-		ServerArgsPerNode: map[int]base.TestServerArgs{
-			0: {
-				StoreSpecs: []base.StoreSpec{
-					{Path: n1s1}, {Path: n1s2}, {Path: n1s3},
-				},
-			},
-			1: {
-				StoreSpecs: []base.StoreSpec{
-					{Path: n2s1}, {Path: n2s2}, {Path: n2s3},
-				},
-			},
-			2: {
-				StoreSpecs: []base.StoreSpec{
-					{Path: n3s1}, {Path: n3s2}, {Path: n3s3},
-				},
-			},
-		},
-	}
+	tcArgs := mkClusterArgs(numNodes, numStoresPerNodeAfterRestart)
+	tcArgs.ReplicationMode = base.ReplicationManual // saves time, ok now
+	// We need ParallelStart since this is an existing cluster. If
+	// we started sequentially, then the first node would hang forever
+	// waiting for the KV layer to become available, but that only
+	// happens when the second node also starts.
+	tcArgs.ParallelStart = true
 
 	// Start all nodes with additional stores.
 	tc = testcluster.StartTestCluster(t, numNodes, tcArgs)

--- a/pkg/server/multi_store_test.go
+++ b/pkg/server/multi_store_test.go
@@ -39,7 +39,10 @@ func TestAddNewStoresToExistingNodes(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	// Nine stores is a lot of goroutines.
 	skip.UnderStress(t, "too many new stores and nodes for stress")
+	skip.UnderRace(t, "too many new stores and nodes for race")
+	skip.UnderDeadlock(t, "too many new stores and nodes for deadlock")
 
 	ctx := context.Background()
 

